doc/qos/new-headroom-calculation.md

```diff
@@ -0,0 +1,168 @@
+
+#### Other tables referenced
+The port speed needs to be fetched from `PORT` table.
+The cable length needs to be fetched from `CABLE_LENGTH` table.
+The calculated headroom size will be stored in `BUFFER_PG` table as it was.
+
+## The detailed flows
+### The well-known formula
+Headroom is calculated as the following:
+ - `headroom` = `Xoff` + `Xon`
+ - `Xon` = `pipeline latency`
+ - `Xoff` = `mtu` + `propagation delay` * `small packet multiply`
+ - `worst case factor` = 2 * `cell` / (1 + `cell`)
+ - `small packet multiply` = (100 - `small packet percentage` + `small packet percentage` * `worst case factor`) / 100
+ - `propagation delay` = `mtu` + 2 * (`kb on cable` + `kb on gearbox`) + `mac/phy delay` + `peer response`
+ - `kb on cable` = `cable length` / `speed of light in media` * `port speed`
+
+The values used in the above procedure are fetched from the following table:
+ - `cable length`: CABLE_LENGTH|\<name\>|\<port\>
+ - `port speed`: PORT|\<port name\>|speed
+ - `kb on gearbox`: SWITCH_PARAMETERS|\<vendor name\>|other_delay
+ - `mac/phy delay`: SWITCH_PARAMETERS|\<vendor name\>|mac_phy_delay
+ - `peer response`: SWITCH_PARAMETERS|\<vendor name\>|peer_response_time
+ - `cell`: SWITCH_PARAMETERS|\<vendor name\>|cell_size
+
+### On start
+
+Try the following steps in order:
+1. Load the table SWITCH_PARAMETERS into an internal dictionary:
+    - If succeeded, set `calculate` flag. Finish.
+    - Otherwise, try 2.
+2. If failed, load the traditional pg_profile_lookup.ini into a lookup table.
+    - If succeeded, set `lookup` flag.
+    - Otherwise, fail the daemon.
+
+### On Port Speed updated
+
+1. Read the speed of the port
+2. Check the following conditions, exit on anyone fails:
+    - Check whether `type` in `BUFFER_PG|<port>|3-4` is of `dynamic` which means dynamically calculating headroom is required for the port.
+    - Check whether there is a cable length configured for the port.
+    - Check whether there is a speed, cable length pair in the internal map.
+3. If there already has been a profile for the given speed, cable length pair in the BUFFER_PROFILE table, just use it.
+4. Otherwise, create a new one by:
+    - If `calculate`, calculating based on formula
+    - If `lookup`, looking up from the lookup table
+   and then insert it into BUFFER_PROFILE table.
+5. Generate a new item referencing the name of the matched/newly-created profile, insert it into BUFFER_PG with key `<Port>|3-4`,
+
+### On Port Cable length updated
+
+1. Check whether the cable length is legal. (Should we support arbitrary cable length?)
+    - If yes, push the length into the internal map.
+    - If no, ignore. Error message should be logged.
+2. After a port's cable length updated, should the BUFFER_PG updated?
+    - If yes, update BUFFER_PG according to its speed and cable length.
+
+## Command line interface
+### To configure a static profile
+```
+sonic#config buffer_profile <name> add <xon> <xoff> <headroom> <dnyamic_th>

```

Comment:
I suggest add more detail about the CLI here, Are all the parameters needed when creating a new profile or we can update part of them afterward?

Reply:
will add.